[ICorProfiler] Add new API to enumerate GC Heap objects

mdh1418 opened the pull request:
This PR aims to add a new ICorProfilerInfo API that extends the ability to enumerate all GC Heap objects (GC's `GCHeap::DiagWalkHeap`) to a profiler. Within a GC there are preparation steps that are performed before+after walking the GC Heap via `DiagWalkHeap`, so this PR introduces a new `IGCHeap` API that facilitates a `DiagWalkHeap` outside of a GC.

Additionally, this PR modifies `ProfToEEInterfaceImpl::SuspendRuntime` + `ProfToEEInterfaceImpl::ResumeRuntime` to both be Asynchronous APIs, invokable directly through the profiler rather than only through callbacks.

-----------
## Testing

### Verification with !dumpheap
Invoking the new EnumerateGCHeapObjects with a walk function like
```
static BOOL STDMETHODCALLTYPE heap_walk_fn(ObjectID object, void* callbackState)
{
    CallbackState* state = static_cast<CallbackState*>(callbackState);
    state->objectCount++;
    ClassID classId{0};
    HRESULT hr = CorProfiler::GetInstance()->pCorProfilerInfo15->GetClassFromObject(object, &classId);
    String typeName = CorProfiler::GetInstance()->GetClassIDName(classId);
    printf("(%ls)\n", typeName.ToWString().c_str());
    return TRUE;
}
```
observed 9999 objects from a sample app.
Attaching windbg to that same sample app and calling `!dumpheap` at the same spot where EnumerateGCHeapObjects was invoked led to 10082 objects, 83 of which were free.

Furthermore, the Class types counts from both the heap walk function and `!dumpheap` matched up.

### Profiler tests
Added unit tests for the following behaviors:
- Custom object type shows up in the GC Heap walk
- Profiler can invoke EnumerateGCHeapObjects within the scope of a profiler requested runtime suspension with SuspendRuntime/ResumeRuntime
- Multiple threads racing to suspend EE will have the slower thread waiting until the faster thread resumes the runtime

mdh1418 committed:
[ICorProfiler] Add new ICorProfilerInfo API to enumerate gc heap

mdh1418 committed:
Update generated corprof.h

After making changes to corprof.idl and building the clr subset
the generated corprof.h needs to be manually copied over from
the artifacts obj directory into the prebuilt directory.

mdh1418 committed:
[ICorProfiler] Extend header and source files to ICorProfilerInfo15

mdh1418 committed:
Cleanup

mdh1418 committed:
[ICorProfiler] Add EnumerateGCHeapObjects skeleton

mdh1418 committed:
[ICorProfiler] Add EnumerateGCHeapObjects implementation

Add parameter check
Add SuspendEE akin to SuspendRuntime
Add DiagWalkHeap
Add RestartEE akin to ResumeRuntime

mdh1418 committed:
[GC] Add IGCHeap API to enumerate GC Heap objects outside of GC

davmason commented on src/coreclr/inc/corprof.idl:
```diff
@@ -4308,6 +4312,33 @@ interface ICorProfilerInfo14 : ICorProfilerInfo13
+     *           it returns S_OK. If it fails, it returns an error code.
+     */
+    HRESULT EnumerateGCHeapObjects(
+        [in] object_callback callback,
```
`object_callback` isn't defined in corprof.idl, I think the build happens to work because it's defined in some other runtime header. We should keep corprof.h self contained as much as possible and provide the definition here. It should be the same definition as `object_callback` but with a new name so it doesn't clash

mdh1418 replied:
I added it in the typedef section at L865, so it is pretty far from where its used. Would it be preferred to move it to right above the interface scope just to keep it close?

I think from a `git grep -w object_callback` in the branch, there doesn't seem to be any other instances besides `corprof.idl`, `corprof.h`, and using it in `proftoeeinterfaceimpl.cpp/.h`

davmason replied:
Oh, I didn't see the definition for some reason. It's fine to leave it where it is but can you change the name to match the style of other callbacks? Something like `ObjectCallback`.

I got it confused with `object_callback_func` in `gcpriv.h

mdh1418 replied:
Changed to ObjectCallback, to be similar casing to the other callback ObjectReferenceCallback

davmason commented on src/coreclr/vm/proftoeeinterfaceimpl.cpp:
```diff
@@ -7615,6 +7619,66 @@ HRESULT ProfToEEInterfaceImpl::GetNonGCHeapBounds(ULONG cObjectRanges,
+
+    // Walk the GC Heap
+    unsigned max_generation = hp->GetMaxGeneration();
+    hp->DiagWalkHeap((walk_fn)callback, callbackState, max_generation, TRUE);
```
@wiktork does the full stop the world gen2 GC meet your performance criteria? I thought you were looking for something lighter weight, although I haven't been super involved in the design process

wiktork replied:
I'm not sure how this compares to dump collection, which is basically what we are trying to be faster than. My previous discussions with Noah suggested that this would be significantly faster.

mdh1418 replied:
For this API we solely suspend EE + fix allocation contexts + walk the gc heap + repair allocation contexts + resume EE, all without performing a GC.

jkotas commented on src/coreclr/gc/gcinterface.h:
```diff
@@ -932,6 +932,9 @@ class IGCHeap {
     virtual void DiagWalkHeap(walk_fn fn, void* context, int gen_number, bool walk_large_object_heap_p) PURE_VIRTUAL
 
+    // Walk the heap object by object outside of a GC.
+    virtual void DiagWalkHeapStandalone(walk_fn fn, void* context, int gen_number, bool walk_large_object_heap_p) PURE_VIRTUAL
```
- Bump the GC interface version https://github.com/dotnet/runtime/blob/c40ecbc6d2fd94a0e9eec5b0154899aa3458dce9/src/coreclr/gc/gcinterface.h#L7-L14

- I think you want to add this method to the very end to make the change non-breaking (minor version bump only)

mdh1418 replied:
Thank you for the catch!

mdh1418 committed:
[GC] Update gc interface minor version

mdh1418 committed:
Match callback PascalCase

noahfalk commented on src/coreclr/pal/prebuilt/inc/corprof.h:
```diff
@@ -55,231 +55,238 @@
 typedef interface ICorProfilerCallback ICorProfilerCallback;
 
-#endif  /* __ICorProfilerCallback_FWD_DEFINED__ */
+#endif 	/* __ICorProfilerCallback_FWD_DEFINED__ */
```
Is there a spaces->tabs conversion that is happening in this file? I'm not sure if you opened the file in your editor which did some automatic whitespace conversion, or something changed in the code generator, or some past checkin accidentally changed the whitespace and this is the first time the file is being regenerated since then.

Not a big deal but it would be nice to confirm that the whitespace that gets checked in is what was produced by the code generator and not an incidental change applied by an auto-formatting editor. It just adds a lot of noise to the diff.

mdh1418 replied:
I believe it is a spaces->tabs conversion when I copied over the file via `cp ./artifacts/obj/coreclr/windows.x64.Debug/inc/idls_out/corprof.h ./src/coreclr/pal/prebuilt/inc/`

I just looked and maybe theres something different between my machine generating the `corprof.h` and @davmason's machine generating the file in #80528?

![Screenshot 2024-06-21 202115](https://github.com/dotnet/runtime/assets/16830051/4040df76-cd3d-4c9f-a0df-d806a0815ad9)

Maybe a configuration setting difference that I might have missed?

jkotas replied:
The build generates the file with tabs (ie what you have in the PR) on my machine as well. My guess is that it was accidental edit that converted tabs to spaces.

davmason replied:
I think I have a setting that converts tabs to spaces automatically, it's probably my fault the file got checked in with spaces

noahfalk replied:
No worries David, we'll convert it back to tabs in this PR and then it should be stable after that.

noahfalk commented on src/coreclr/vm/proftoeeinterfaceimpl.cpp:
```diff
@@ -7615,6 +7619,57 @@ HRESULT ProfToEEInterfaceImpl::GetNonGCHeapBounds(ULONG cObjectRanges,
+{
+    CONTRACTL
+    {
+        // Yay!
```
Welcome to David Broman's interesting commenting convention ;p
(I'm not suggesting any change)

noahfalk reviewed (commented):
This looks good so far, I think the main thing remaining is to add a test case for the new API.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -52052,6 +52052,34 @@ void GCHeap::DiagWalkHeap (walk_fn fn, void* context, int gen_number, bool walk_
     gc_heap::walk_heap (fn, context, gen_number, walk_large_object_heap_p);
 }
 
+void GCHeap::DiagWalkHeapStandalone (walk_fn fn, void* context, int gen_number, bool walk_large_object_heap_p)
```
as I mentioned when we chatted, I was hoping we'd come up with a name better than Standalone, especially now since we don't do suspend/restart EE anymore in this, maybe just call it something like DiagWalkHeapWithACHandling. feel free to suggest something better.

also I would add a comment that says the caller is responsible for suspending the EE before callling this.

noahfalk replied:
I thought about it but I didn't come up with any name suggestion that felt better. One convention used elsewhere in profiler code is that we call methods "Async" when they can be called at arbitrary times vs. "Synchronous" when they can only be called from within profiler callbacks. Its overloaded terminology but applies here if you wanted to use it.

I'm just fine calling this DiagWalkHeapWithACHandling or any other name folks would like.

jkotas replied:
What are the chances that you will want to add another flag to this method? Would it make sense for this method to have more general name and take flags to control the specific behavior?

```
typedef enum
{
    walk_heap_large_object_heap = 1,
    walk_heap_update_allocation_contexts = 2,
} walk_heap_type;
```

```
void GCHeap::DiagWalkHeap2 (walk_fn fn, void* context, int gen_number, walk_heap_type type)
```

mdh1418 replied:
> What are the chances that you will want to add another flag to this method?

I'm thinking that it is low if there aren't any plans to add more behaviors for how the gc heap is walked
The way I see it, this method spans a larger scope of the 5 steps needed to walk the gc heap compared to `GCHeap::DiagWalkHeap` which is solely called within a GC and assumes the other steps have/will happen.
```
Suspending EE
    Fix all heap allocation contexts            | (This method's scope)
        Walking GC Heap (DiagWalkHeap's scope)  | 
    Repair all heap allocation contexts         |
Resuming EE
```
There was the thought of including the EE suspend/resume in this method's scope, but initially I felt that leaving the responsibility of EE suspend/resume to the ICorProfiler API was simpler because:
1. all other EE suspensions in gc.cpp are [GCToEEInterface::SuspendEE](https://github.com/dotnet/runtime/blob/c241cc99ac0cb3862bdee5e6f931f0ba60537631/src/coreclr/vm/gcenv.ee.cpp#L38-L52), whose assertions aren't appropriate since this method doesn't actually perform a GC
2. Other ICorProfiler API's already [suspend EE](https://github.com/dotnet/runtime/blob/c241cc99ac0cb3862bdee5e6f931f0ba60537631/src/coreclr/vm/proftoeeinterfaceimpl.cpp#L6837-L6868)/[resume EE ](https://github.com/dotnet/runtime/blob/c241cc99ac0cb3862bdee5e6f931f0ba60537631/src/coreclr/vm/proftoeeinterfaceimpl.cpp#L6870-L6900), so there's precedence + logic accounting for an already suspended EE makes sense.

Re: Naming, I also didn't come up with any better names (`DiagWalkHeapOutsideOfGC` didn't feel better than `DiagWalkHeapStandalone`), but `DiagWalkHeapWithACHandling` seems more appropriate especially considering the scoping above.

I'll add comments that EE Suspension/Resumption responsibility falls to callers of `GCHeap::DiagWalkHeapWithACHandling`.

mdh1418 committed:
Address feedback - naming and description

mdh1418 committed:
Cleanup

mdh1418 committed:
Add EEToProf EnumerateGCheapObjectsCallback wrapper

Profilers may want to inspect objects encountered through the callback
with synchronous ICorProfilerInfo APIs. Wrapping the GC Heap walk
within a EEToProfInterfaceImpl helper with a CLR_TO_PROFILER_ENTRYPOINT
macro allows proper setting of callback state flags.

jkotas commented on src/coreclr/inc/profilepriv.inl:
```diff
@@ -1762,6 +1762,23 @@ inline void ProfControlBlock::EventPipeProviderCreated(EventPipeProvider *provid
+{
+    LIMITED_METHOD_CONTRACT;
+
+    return DoProfilerCallback(ProfilerCallbackType::Active,
```
This will iterate all attached profilers and invoke the callback for each one of them. I do not think it is what we want for the `EnumerateGCHeapObjects` API. The callback should be only called for the one profiler that called the API.

jkotas commented on src/coreclr/vm/proftoeeinterfaceimpl.cpp:
```diff
@@ -7615,6 +7619,64 @@ HRESULT ProfToEEInterfaceImpl::GetNonGCHeapBounds(ULONG cObjectRanges,
+    // invoke ICorProfilerInfo APIs guarded by AllowObjectInspection by toggling fGCInProgress.
+    g_profControlBlock.fGCInProgress = TRUE;
+
+    // Leveraging a direct callback instead of a ICorProfilerCallback API avoids the performance overhead of
```
I do not see what we are trying to protect against here. The profiler should prevent itself from getting unloaded when it calls EnumerateGCHeapObjects. It is not the runtime job to prevent the profiler from getting unloaded in this situation.

davmason replied:
We don't need to guard against unloading, we need to set the profiler thread flags so the profiler can call the `PROFILER_TO_CLR_ENTRYPOINT_SYNC` methods and do object inspection in the callback

jkotas replied:
Yes, your description makes sense.

This calls `DoProfilerCallback`. `DoProfilerCallback` is mostly concerned with calling all attached profilers that is not something that we want to do here.

Is there a better - more direct and lightweight - way to set the profiler flags?

mdh1418 replied:
There might be a way through `SetProfilerCallbackStateFlags`, since https://github.com/dotnet/runtime/blob/6b4b371c44ea14536f860b75b5a1553605f49c82/src/coreclr/vm/proftoeeinterfaceimpl.cpp#L244 is really what is blocking the non-EEToProfInterfaceImpl callback from invoking synchronous APIs guarded with the logic under the`PROFILER_TO_CLR_ENTRYPOINT_SYNC_EX` macro. But didn't go down that route, and opted to use a "shell" EEToProfInterfaceImpl wrapper helper.

jkotas replied:
The wrapper is fine. The part that I do not understand is what you get by going through `DoProfilerCallback`.

What is not going to work if you call EnumerateGCHeapObjectsCallback directly from here instead of going through `DoProfilerCallback`?

mdh1418 replied:
I originally mimicked the pattern to invoke EEToProfInterfaceImpl method, not realizing at the time that DoProfilerCallback would iterate over all attached profilers.

Right now, given that the macro for `PROFILER_TO_CLR_ENTRYPOINT_SYNC_EX` uses members of the EEToProfIntefaceImpl class + `this`, I wouldn't be able to invoke it statically. I didn't see (from a glance) any instances of calling EEToProfInterfaceImpl methods from ProfToEEInterfaceImpl APIs that weren't calling through `g_profControlBlock`, so I naively followed suit there at the time.

What would be the best way to grab the "current" profiler? I'm wondering if something like ` mainProfilerInfo.pProfInterface.Load()` makes sense, or if there would ever be an instance where a non-main profiler would invoke `EnumerateGCHeapObjects`.

jkotas replied:
> What would be the best way to grab the "current" profiler?

Can you get it from `ProfToEEInterfaceImpl::m_pProfilerInfo`?

mdh1418 replied:
Used m_pProfilerInfo to grab the EEToProfInterfaceImpl and invoked the callback `EnumerateGCHeapObjectsCallback` directly

jkotas commented on src/coreclr/inc/profilepriv.inl:
```diff
@@ -1762,6 +1762,16 @@ inline void ProfControlBlock::EventPipeProviderCreated(EventPipeProvider *provid
+// and instead, invokes the callback directly on the EEToProfInterfaceImpl instance.
+inline HRESULT ProfControlBlock::EnumerateGCHeapObjectsCallback(EEToProfInterfaceImpl *profInterface, ObjectCallback callback, void* callbackState)
+{
+    return profInterface->EnumerateGCHeapObjectsCallback(callback, callbackState);
```
Would it be better to make this call directly from `ProfToEEInterfaceImpl::EnumerateGCHeapObjects` and delete this method?

mdh1418 replied:
Ah right, since we can grab the EEToprofInterfaceImpl directly, we don't need to route through `g_profControlBlock`. Thanks for the catch!

mdh1418 replied:
Removed the ProfControlBlock changes

davmason commented on src/coreclr/vm/proftoeeinterfaceimpl.cpp:
```diff
@@ -7615,6 +7619,74 @@ HRESULT ProfToEEInterfaceImpl::GetNonGCHeapBounds(ULONG cObjectRanges,
+        return CORPROF_E_RUNTIME_UNINITIALIZED;
+    }
+
+    if (!ThreadSuspend::SysIsSuspendInProgress() && (ThreadSuspend::GetSuspensionThread() == 0))
```
Are you trying to only suspend if the profiler has not already suspended the runtime? I worry about this check because it doesn't distinguish between a profiler induced suspension and a random suspension for some other reason.

Also, the `RestartEE` call below isn't in an if statement so if this does not suspend the runtime, it will try to restart the runtime in an unbalanced way.

Is there a reason we can't just unconditionally suspend the runtime here?

mdh1418 replied:
I added logic to suspend/resume EE in the same manner that `ProfToEEInterfaceImpl::SuspendRuntime` and `ProfToEEInterfaceImpl::ResumeRuntime` do it. I haven't yet delved into the details of whether a double `ThreadSuspend::SuspendEE` before a restart would cause any issues, @noahfalk/@jkotas would y'all happen to know? 

Should this API not proceed if there was a non-profiler induced suspension (since whatever suspended EE might try to resume EE before this API finishes)? We could check for a previous profiler-induced suspension with `fProfilerRequestedRuntimeSuspend`, but then a similar question pops up, whether that previous profiler-induced suspension is already paired with a resumption.

Would it make sense to require profilers to handle EE Suspension/Resumption outside of this EnumerateGCHeapObjects API rather than building it into the logic?
e.g.
```cpp
pCorProfilerInfo->SuspendRuntime();
...
pCorProfilerInfo->EnumerateGCHeapObjects(heap_walk_fn, &state);
...
pCorProfilerInfo->ResumeRuntime();
```

jkotas replied:
> double ThreadSuspend::SuspendEE

I do not think double ThreadSuspend::SuspendEE is going to work correctly. You can prove or disprove it by adding tests.

mdh1418 replied:
Both double `ThreadSuspend::SuspendEE` and `ThreadSuspend::RestartEE` will fail assertions, added logic to only invoke `ThreadSuspend::RestartEE` in this API if `ThreadSuspend::SuspendEE` had been invoked in the API.

jkotas replied:
Give that this API has `EE_THREAD_NOT_REQUIRED` contract, it is expected to work on background threads that have not joined the runtime. If this is called on a background thread, there may be suspension running on the foreground thread, `ThreadSuspend::SysIsSuspendInProgress()` returns true and we will skip our own suspension here. However, nothing guarantees that the suspension that was performed by the foreground thread is going to last as long as this method runs. In extreme case, the runtime may be resumed and running by the time we hit the next line that sets `g_profControlBlock.fGCInProgress = TRUE;`.

I think the contract of the method needs to be made more stringent to avoid this situation; or the implementation should be fixed to handle this situation.

noahfalk replied:
I've been answering some questions in other communication channels piece-meal but since it still sounded like there was ongoing uncertainty I wanted to offer a more definitive suggestion.

I think this would give us good behavior:
```
bool ownEESuspension = false;
bool suspendedByThisThread = (ThreadSuspend::GetSuspensionThread() == GetThreadNullOK());
if (suspendedByThisThread && !g_profControlBlock.fProfilerRequestedRuntimeSuspend)
{
    // This thread is responsible for suspending the runtime so we can't block
    // waiting for the runtime to resume. However it wasn't the profiler call that did
    // the suspend so we don't know what state the GC heap is in. Other threads also might 
    // be modifying it concurrently. In the future more analysis or coordination with other
    // suspenders might let us narrow the scope of this error condition, but we have no need
    // to do this now.
    return CORPROF_E_SUSPENSION_IN_PROGRESS;
}
else if(suspendedByThisThread && g_profControlBlock.fProfilerRequestedRuntimeSuspend)
{
    // This thread previously invoked ICorProfiler::SuspendRuntime(). Our caller
    // has the responsibility to resume the runtime no earlier than when this API returns
    // and to preserve the GC heap in a consistent state. We should avoid invoking 
    // SuspendEE/ResumeEE again in this function because those APIs do not support 
    // re-entrant suspends.
}
else
{
    _ASSERTE(!suspendedByThisThread);
    // Its possible some background threads are suspending and resuming the runtime
    // concurrently. We need to suspend the runtime on this thread to be certain the heap
    // stays in a walkable state for the duration that we need it to. Our call to
    // SuspendEE() may race with other threads by design and this thread may block
    // arbitrarily long inside SuspendEE() for other threads to complete their own
    // suspensions.
    g_profControlBlock.fProfilerRequestedRuntimeSuspend = TRUE;
    ThreadSuspend::SuspendEE(ThreadSuspend::SUSPEND_REASON::SUSPEND_FOR_PROFILER);
    ownEESuspension = TRUE;
}
```

mdh1418 replied:
Thanks for the suggestion! Added it

davmason reviewed (approved):
Other than the one comment around suspension that we need to settle on the rest looks good to me.

mdh1418 committed:
[Profiler][Tests] Add EnumerateGCHeapObjects unit test

mdh1418 committed:
Address feedback

As EE may be suspended outside of ProfToEEInterfaceImpl::EnumerateGCHeapObjects,
conditionally suspend/resume EE if no prior suspension is in progress.

mdh1418 committed:
[ICorProfiler] Update Callback helper contract

As EnumerateGCHeapObjectsCallback should only be invoked by
ProfToEEInterfaceImpl::EnumerateGCHeapObjects which suspends EE,
signal that no other EE suspension should occur with GC_NOTRIGGER
and kEE2PNoTrigger.

Inherit MODE_ANY from EnumerateGCHeapObjects

mdh1418 committed:
[ICorProfiler] Change Suspend/Resume Runtime to Asynchronous

mdh1418 committed:
[ICorProfiler] Mitigate runtime suspension racing with EnumerateGCHeapObjects

As the runtime may be suspended in other ways, error if not a profiler
requested suspension similar to ProfToEEInterfaceImpl::RuntimeSuspend

mdh1418 committed:
Update test cases

mdh1418 committed:
[Tests] Update profiler requested runtime test

mdh1418 committed:
fixup

mdh1418 committed:
Add test for background EnumerateGCHeapObject

mdh1418 committed:
Set profiler requested runtime suspend flag

mdh1418 committed:
Fix symbol exports

mdh1418 committed:
Try fix compiler issues

tommcdon commented on src/coreclr/vm/proftoeeinterfaceimpl.cpp:
```diff
@@ -7615,6 +7619,91 @@ HRESULT ProfToEEInterfaceImpl::GetNonGCHeapBounds(ULONG cObjectRanges,
+        EEToProfInterfaceImpl *pProfInterface = m_pProfilerInfo->pProfInterface.Load();
+        if (pProfInterface == NULL)
+        {
+            return E_FAIL;
```
If `ownEESuspension == TRUE` then should we call `ThreadSuspend::RestartEE` before returning?

mdh1418 replied:
Yes, good catch.

mdh1418 committed:
Allow runtime to resume before returning

mdh1418 committed:
Fix compilation errors on non-windows platforms

mdh1418 committed:
Lower expected object count

mdh1418 committed:
Add profiler requested runtime suspension note

mdh1418 committed:
[Tests] Protect custom test object from GC

mdh1418 commented:
Still having a bit of trouble reproducing the CI failures, tried to follow the build steps through the different legs via
[./build.cmd -arch x64 -os windows  -s clr+libs -c Debug -rc Checked](https://dev.azure.com/dnceng-public/public/_build/results?buildId=733479&view=logs&j=45eeb76f-ea8f-5323-0629-ae4135794e72&t=c7572688-1f09-5a0d-1591-da67cfb1acd3&l=11)
[./src/tests/build.cmd skipmanaged skipgeneratelayout Checked x64](https://dev.azure.com/dnceng-public/public/_build/results?buildId=733479&view=logs&j=45eeb76f-ea8f-5323-0629-ae4135794e72&t=66504fd8-803b-5730-d5ad-51ac1846a45f&l=11)
[./src/tests/build.cmd copynativeonly log Native    checked x64   /p:LibrariesConfiguration=Debug](https://dev.azure.com/dnceng-public/public/_build/results?buildId=733479&view=logs&j=36023878-3ad4-5624-c1ed-575c64898e0b&t=6d235163-eb2d-500c-52fb-5df52e1aea3c&l=11)
[./src/tests/build.cmd buildtestwrappersonly log Wrappers   checked x64    /p:LibrariesConfiguration=Debug](https://dev.azure.com/dnceng-public/public/_build/results?buildId=733479&view=logs&j=36023878-3ad4-5624-c1ed-575c64898e0b&t=ee8a32c0-d4bc-50ec-a9c7-de856b414ced&l=11)
[./src/tests/build.cmd generatelayoutonly log Layout   checked x64    /p:LibrariesConfiguration=Debug](https://dev.azure.com/dnceng-public/public/_build/results?buildId=733479&view=logs&j=36023878-3ad4-5624-c1ed-575c64898e0b&t=369775f3-f8c8-57d7-7a4b-c2acfeb9f674&l=11)
Maybe I'm missing an intermediate step, but I attempt to run the `dotnet` command shown in the [console log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-103735-merge-ee598f32a34d45eab4/profiler.gcheapenumeration/1/console.d1b5af53.log?helixlogtype=result) but I get 
```
BEGIN: coreclr_initialize failed - Error: 0x80070057
END: coreclr_initialize failed - Error: 0x80070057
```

and I think when I attempted to run the test command script directly `.\artifacts\tests\coreclr\windows.x64.Checked\profiler\gcheapenumeration\gcheapenumeration\gcheapenumeration.cmd` there wasn't a `gcheapenumeration.dll` until I built the test manually with ` .\src\tests\build.cmd test profiler/gcheapenumeration/gcheapenumeration.csproj /p:Configuration=Checked /p:LibrariesConfiguration=Debug` which seems to deviate from how its built on CI.

And building manually/running the test command, I haven't been able to repro the error on CI, which seems to be unable to find the custom GCHeap object for all checked no_tiered_compilation test legs.

Any other ideas to repro @noahfalk ? Where might the dumps be for the windows/linux runtime tests legs?
I can also continue to try to push up changes to the tests just to see if our suspected fixes might work.

noahfalk commented on src/coreclr/vm/proftoeeinterfaceimpl.cpp:
```diff
@@ -7615,6 +7619,95 @@ HRESULT ProfToEEInterfaceImpl::GetNonGCHeapBounds(ULONG cObjectRanges,
+    }
+    CONTRACTL_END;
+
+    PROFILER_TO_CLR_ENTRYPOINT_ASYNC_EX(kP2EEAllowableAfterAttach | kP2EETriggers,
```
I noticed, somewhat surprisingly, that the implementation of the ASYNC macro doesn't do any validation based on the kP2EETriggers flag. While we might want to improve the macro itself in the future, given we are near the end of feature development and its not causing known problems its probably safer to add a comment and some additional validation local to this method. Something like:

```
PROFILER_TO_CLR_ENTRYPOINT_ASYNC_EX(...);
if(!IsCalledAsynchronously() && !AreCallbackStateFlagsSet(COR_PRF_CALLBACKSTATE_IN_TRIGGERS_SCOPE))
{
    LOG((LF_CORPROF, LL_ERROR,
                 "**PROF: ERROR: Returning CORPROF_E_UNSUPPORTED_CALL_SEQUENCE "
                 "due to illegal asynchronous profiler call\n"));
    return CORPROF_E_UNSUPPORTED_CALL_SEQUENCE;
}
```

This should catch most misuse though it wouldn't detect complicated cases like a profiler that hijacks a thread running managed code and then tries to do enumeration from inside the hijack. I'm sure we could make a stronger check if we needed to, but I think this would be good for now.

mdh1418 replied:
Good catch, added the extra check to SuspendRuntime, ResumeRuntime, and EnumerateGCHeapObjects

noahfalk commented on src/coreclr/vm/proftoeeinterfaceimpl.cpp:
```diff
@@ -6846,7 +6850,7 @@ HRESULT ProfToEEInterfaceImpl::SuspendRuntime()
     CONTRACTL_END;
 
-    PROFILER_TO_CLR_ENTRYPOINT_SYNC_EX(
+    PROFILER_TO_CLR_ENTRYPOINT_ASYNC_EX(
```
Since the ASYNC macro doesn't give us the same error checking on the k2PEETriggers flag it would be good to add the same check I recommended in EnumerateGCHeap objects here.

mdh1418 replied:
Done

noahfalk commented on src/tests/profiler/native/gcheapenumerationprofiler/gcheapenumerationprofiler.cpp:
```diff
@@ -0,0 +1,237 @@
+    HRESULT hr = state->instance->pCorProfilerInfo->GetClassFromObject(object, &classId);
+    if (hr != S_OK)
+    {
+        printf("Error: failed to get class ID from object.\n");
```
I think we should expect that every object has an obtainable class, if not the test should fail. This prints out a console error message but it doesn't appear to record anything that makes the test fail.

mdh1418 replied:
Made the failure counter publicly incrementable, and changed the walk function to stop enumerating the GC Heap if a class could not be resolved.

noahfalk commented on src/tests/profiler/gcheapenumeration/gcheapenumeration.cs:
```diff
@@ -0,0 +1,144 @@
+        public static int EnumerateGCHeapObjectsSingleThreadWithinProfilerRequestedRuntimeSuspension()
+        {
+            _rootObjects.Add(new CustomGCHeapObject());
+            SuspendRuntime();
```
This doesn't look like it would work. If this SuspendRuntime() returns back to managed code that means the runtime isn't suspended. This seems like it either deadlocks or it isn't suspending.

I'd recommend making a native method that invokes all of these functions rather than trying to coordinate it from managed code.

noahfalk replied:
Its a surprising result. My best guess is that some optimization in how the JIT does p/invokes causes it to optimize away barrier that prevents returning to managed code in between adjacently called p/invokes. I would not recommend writing the test in a way that relies on that kind of optimization though. I'd still suggest writing a single native method that does Suspend(), Enumerate(), and Resume() clearly within native code.

mdh1418 replied:
Added `EnumerateGCHeapObjectsWithinProfilerRequestedRuntimeSuspension` which does all three

noahfalk commented on src/tests/profiler/gcheapenumeration/gcheapenumeration.cs:
```diff
@@ -0,0 +1,144 @@
+            Thread enumerateThread = new Thread(() =>
+            {
+                startEvent.WaitOne();
+                Thread.Sleep(1000);
```
I don't think this is testing what you intend it to. The GC.Collect() keeps managed code suspended which means this sleep isn't waking up until the GC.Collect() is finished. By the time EnumerateGCHeapObjects() runs there is no concurrent suspension happening. 

To test concurrent suspends you will need to coordinate it from native code. For example you might do this:
```
GCHeapEnumerationProfiler::GarbageCollectionStarted()
{
    _gcStartSleeping = TRUE;
    // if this new call to EnumerateGCHeapObjects doesn't wait for the GC then it is going to
    // observe _gcStartSleeping == TRUE
    thread_list.emplace_back(std::thread([instance]()
                {
                    instance->ValidateEnumerateGCHeapObjects(S_OK);
                }));
   std::this_thread::sleep_for(std::chrono::seconds(1));
   _gcStartSleeping = FALSE;
}

static BOOL STDMETHODCALLTYPE heap_walk_fn(ObjectID object, void* callbackState)
{
    CallbackState* pState = (CallbackState*)callbackState;
    if(pState->instance._gcStartSleeping)
    {
        //Error, no callbacks expected while the GC collection was going on
    }
    ...
}

HRESULT GCHeapEnumerationProfiler::Shutdown()
{
    // make sure async operations have finished
    for (auto& t : threadList) {
        t.join();
    }
}
```

mdh1418 replied:
For clarification, this would be testing invoking `ProfToEEInterfaceImpl::EnumerateGCHeapObjects` during a non-profiler requested runtime suspension? As in we are expecting it to hit the `else if` in 
```
    if (!ThreadSuspend::SysIsSuspendInProgress() && (ThreadSuspend::GetSuspensionThread() == 0))
    {
        g_profControlBlock.fProfilerRequestedRuntimeSuspend = TRUE;
        ThreadSuspend::SuspendEE(ThreadSuspend::SUSPEND_REASON::SUSPEND_FOR_PROFILER);
        ownEESuspension = TRUE;
    }
    else if (!g_profControlBlock.fProfilerRequestedRuntimeSuspend)
    {
        return CORPROF_E_SUSPENSION_IN_PROGRESS;
    }
    else
    {
        // The profiler requested a runtime suspension before invoking this API,
        // so the responsibility of resuming the runtime is outside the scope of this API.
        // Given that the profiler owns the suspension, walking the GC Heap is safe.
    }
```
and throw the error `CORPROF_E_SUSPENSION_IN_PROGRESS`

Originally, I was intending the [3rd test](https://gist.github.com/mdh1418/3bb38bbb120ab99e6c3f2225aaca041f)
to be the "background native thread invoking EnumerateGCHeapObjects and rejoining a thread that has an ongoing non-profiler requested runtime suspension" test, and for this (4th test) to be multiple managed threads racing for `ThreadSuspend::SuspendEE`, in case the above if block
```
    if (!ThreadSuspend::SysIsSuspendInProgress() && (ThreadSuspend::GetSuspensionThread() == 0))
    {
        g_profControlBlock.fProfilerRequestedRuntimeSuspend = TRUE;
        ThreadSuspend::SuspendEE(ThreadSuspend::SUSPEND_REASON::SUSPEND_FOR_PROFILER);
        ownEESuspension = TRUE;
    }
```
is entered before a different native thread invokes `ThreadSuspend::SuspendEE` (which would have toggled the condition to be false)

noahfalk replied:
> As in we are expecting it to hit ProfToEEInterfaceImpl::EnumerateGCHeapObjects's else if in

I proposed some alternate logic for how you would check for suspensions. The test case I suggested in my comment above does the EnumerateGCHeapObjects on a different thread than the one which suspended the runtime so I'd expect it to enumerate successfully after waiting for the GC to complete. It should run through this part of the code I recommended above:

```
else
{
    _ASSERTE(!suspendedByThisThread);
    // Its possible some background threads are suspending and resuming the runtime
    // concurrently. We need to suspend the runtime on this thread to be certain the heap
```

mdh1418 replied:
Ah I missed the above recommendations for checking thread suspension within `ProfToEEInterfaceImpl::EnumerateGCHeapObjects`, after implementing a test as you suggested in the first comment in this thread, I attached a debugger and the background thread that invokes EnumerateGCHeapObjects does indeed get blocked on `ThreadSuspend::SuspendEE` until GC completes
```
00 0000001b`0657f218 00007ffb`06d640e3     ntdll!ZwWaitForAlertByThreadId+0x14
01 0000001b`0657f220 00007ffb`06d51ce4     ntdll!TpWorkOnBehalfClearTicket+0x2c3
02 0000001b`0657f330 00007ffb`06d51ad2     ntdll!RtlEnterCriticalSection+0x254
03 0000001b`0657f3b0 00007ffa`5b26652e     ntdll!RtlEnterCriticalSection+0x42
04 0000001b`0657f3e0 00007ffa`5b51372b     coreclr!CrstBase::Enter+0x27e [C:\Users\mihw\source\repos\mdh1418\runtime\src\coreclr\vm\crst.cpp @ 325] 
05 0000001b`0657f440 00007ffa`5b8abd8b     coreclr!ThreadStore::Enter+0x1bb [C:\Users\mihw\source\repos\mdh1418\runtime\src\coreclr\vm\threads.cpp @ 5092] 
06 0000001b`0657f560 00007ffa`5b8b18ad     coreclr!ThreadSuspend::LockThreadStore+0x25b [C:\Users\mihw\source\repos\mdh1418\runtime\src\coreclr\vm\threadsuspend.cpp @ 1882] 
07 0000001b`0657f690 00007ffa`5b808f13     coreclr!ThreadSuspend::SuspendEE+0x12d [C:\Users\mihw\source\repos\mdh1418\runtime\src\coreclr\vm\threadsuspend.cpp @ 5559] 
08 0000001b`0657f770 00007ffa`672cd542     coreclr!ProfToEEInterfaceImpl::EnumerateGCHeapObjects+0x473 [C:\Users\mihw\source\repos\mdh1418\runtime\src\coreclr\vm\proftoeeinterfaceimpl.cpp @ 7707] 
09 0000001b`0657f8f0 00007ffa`672cdda8     Profiler!GCHeapEnumerationProfiler::EnumerateGCHeapObjects+0xa2 [C:\Users\mihw\source\repos\mdh1418\runtime\src\tests\profiler\native\gcheapenumerationprofiler\gcheapenumerationprofiler.cpp @ 186] 
0a 0000001b`0657f960 00007ffa`672cd335     Profiler!GCHeapEnumerationProfiler::ValidateEnumerateGCHeapObjects+0x18 [C:\Users\mihw\source\repos\mdh1418\runtime\src\tests\profiler\native\gcheapenumerationprofiler\gcheapenumerationprofiler.cpp @ 190] 
0b 0000001b`0657f9a0 00007ffa`672ccab4     Profiler!<lambda_e2419102f21665b3332b03f113586c25>::operator()+0x25 [C:\Users\mihw\source\repos\mdh1418\runtime\src\tests\profiler\native\gcheapenumerationprofiler\gcheapenumerationprofiler.cpp @ 74] 
0c 0000001b`0657f9d0 00007ffa`672cbe19     Profiler!std::invoke<<lambda_e2419102f21665b3332b03f113586c25> >+0x14 [C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.40.33807\include\type_traits @ 1705] 
0d 0000001b`0657fa00 00007ffa`673b63a0     Profiler!std::thread::_Invoke<std::tuple<<lambda_e2419102f21665b3332b03f113586c25> >,0>+0x59 [C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.40.33807\include\thread @ 61] 
0e 0000001b`0657fa60 00007ffb`055f257d     Profiler!thread_start<unsigned int (__cdecl*)(void *),1>+0xb0 [minkernel\crts\ucrt\src\appcrt\startup\thread.cpp @ 97] 
0f 0000001b`0657fac0 00007ffb`06d8af28     KERNEL32!BaseThreadInitThunk+0x1d
10 0000001b`0657faf0 00000000`00000000     ntdll!RtlUserThreadStart+0x28
```

mdh1418 replied:
Modified the test case to simulate scenario 1, where the thread suspending runtime through GC spins off another native thread to invoke EnumerateGCHeapObjects.

Will add the test case to simulate scenario 2 if I can find the test attribute that bypasses contract violation assert and hits the expected errors.

noahfalk commented on src/tests/profiler/native/gcheapenumerationprofiler/gcheapenumerationprofiler.cpp:
```diff
@@ -0,0 +1,237 @@
+    SHUTDOWNGUARD();
+    printf("GCHeapEnumerationProfiler::GarbageCollectionStarted\nSleeping for 10 seconds\n");
+    std::this_thread::sleep_for(std::chrono::seconds(10));
+    return S_OK;
```
Lets validate that if you try to call EnumerateGCHeapObjects from inside CollectionStart/CollectionFinished that it doesn't work. I believe these callbacks are both no-trigger callbacks so I'd expect the calls to fail with CORPROF_E_UNSUPPORTED_CALL_SEQUENCE if you've added the additional kP2EETriggers check above.

mdh1418 replied:
Calling EnumerateGCHeapObjects from inside GarbageCollectionStarted throws a Contract violation assert from the `CONTRACTL/CONTRACTL_END` scope before hitting the `PROFILER_TO_CLR_ENTRYPOINT_ASYNC_EX` macro and additional `!IsCalledAsynchronously() && !AreCallbackStateFlagsSet(COR_PRF_CALLBACKSTATE_IN_TRIGGERS_SCOPE)` check

noahfalk replied:
yeah I expect that test case would only work in a retail build. I thought we had some test attribute that did that but don't worry about including it if it isn't straigtforward.

noahfalk commented on src/tests/profiler/native/gcheapenumerationprofiler/gcheapenumerationprofiler.cpp:
```diff
@@ -0,0 +1,237 @@
+    {
+        state->customGcHeapObjectTypeCount->fetch_add(1, std::memory_order_relaxed);
+    }
+
```
Lets verify that you can't re-entrantly call EnumerateGCHeapObjects from inside the object callback. I'd expect CORPROF_E_UNSUPPORTED_CALL_SEQUENCE.

mdh1418 replied:
Like in https://github.com/dotnet/runtime/pull/103735/files#r1671330380, its hitting a contract violation first from calling a GC_TRIGGERS within a GC_NOTRIGGER (from the callback EEToProfInterfaceImpl::EnumerateGCHeapObjectsCallback)

noahfalk commented on src/tests/profiler/native/gcheapenumerationprofiler/gcheapenumerationprofiler.cpp:
```diff
@@ -0,0 +1,237 @@
+                {
+                    std::this_thread::sleep_for(std::chrono::seconds(1));
+                    instance->ValidateEnumerateGCHeapObjects(0x80131388);
+                }).detach();
```
If the test creates threads I'd keep track of them and make sure to join them later. Otherwise there is no guarantee the code being tested on these threads has ran prior to the test reporting its PASS/FAIL results. There is also the possibility that calling profiler API asynchronously during runtime shutdown/process exit could generate unexpected error conditions.

mdh1418 replied:
Added the suggested logic at the start of Shutdown to wait for all threads tracked to finish execution with join().

mdh1418 committed:
Fix Profiler Runtime Suspension test

mdh1418 committed:
Expect to resolve classes during heap walk

mdh1418 committed:
Address feedback

mdh1418 commented on src/tests/profiler/gcheapenumeration/gcheapenumeration.cs:
```diff
@@ -0,0 +1,144 @@
+                                              profileeArguments: testName,
+                                              envVars: new Dictionary<string, string>
+                                              {
+                                                {ShouldSetMonitorGCEventMaskEnvVar, shouldSetMonitorGCEventMask},
```
Changed to bool

wiktork commented on src/tests/profiler/native/gcheapenumerationprofiler/gcheapenumerationprofiler.cpp:
```diff
@@ -0,0 +1,261 @@
+HRESULT STDMETHODCALLTYPE GCHeapEnumerationProfiler::GarbageCollectionStarted(int cGenerations, BOOL generationCollected[], COR_PRF_GC_REASON reason)
+{
+    SHUTDOWNGUARD();
+    printf("GCHeapEnumerationProfiler::GarbageCollectionStarted\n");
```
Is the choice to call EnumerateGCHeapObjects from GarbageCollectionStarted simply an easy way to test? Or can we call it at any point as long as we call SuspendRuntime first?

mdh1418 replied:
EnumerateGCHeapObjects will handle EE suspension, and as it is an asynchronous API, it can be both directly called by the profiler, as well as invoked by a callback (demonstrated here). The `EnumerateGCHeapObjectsWithoutProfilerRequestedRuntimeSuspension` PInvoke demonstrates invoking the API directly from the profiler.

Calling EnumerateGCHeapObjects from GarbageCollectionStarted was simply to test the behavior when a different native thread has suspended EE (EnumerateGCHeapObjectsMultiThreadWithCompetingRuntimeSuspension test case). In this case, the current thread has suspended EE through GC, and it spins off a separate native thread to EnumerateGCHeapObjects, demonstrating that ThreadSuspend::SuspendEE in the other thread will block until the native thread that owns suspension has RestartedEE.

wiktork commented on src/tests/profiler/native/gcheapenumerationprofiler/gcheapenumerationprofiler.cpp:
```diff
@@ -0,0 +1,261 @@
+        state->instance->IncrementFailures();
+        return FALSE;
+    }
+
```
Can we also validate that we can get the class layout here (using GetClassLayout) since ultimately we'll want to read fields on these types?

mdh1418 replied:
I'm able to call that API as well from the heap walk fn, and for a majority of the classId's I'm hitting these Invalid args https://github.com/dotnet/runtime/blob/041375910885f3a30ccf0bb8b751566dae7043f4/src/coreclr/vm/proftoeeinterfaceimpl.cpp#L7848-L7860. Also not quite sure yet how to extract the information from `COR_FIELD_OFFSET` to get useful field info

noahfalk replied:
These layout APIs expect you to divide up heap objects into strings, arrays, and everything else. GetClassLayout() only returns a successful result on the 'everything else' category. To determine if a type is an array you can call ICorProfilerInfo::IsArrayClass(). To identify string you can check if the type name is "System.String". A more efficient profiler (ie not test code) would probably want to cache the unique ClassID value for string and then do a direct comparison on the ClassID to identify strings.

noahfalk replied:
> Also not quite sure yet how to extract the information from COR_FIELD_OFFSET to get useful field info

COR_FIELD_OFFSET has a FieldDef metadata token you can look up with metadata APIs to get a name. It also has an offset which represents the memory offset from the start of the object where data for that field is stored. I think its sufficient just to verify that GetClassLayout returned S_OK and assume the layout information it is giving you is correct.

mdh1418 replied:
Added validation to check that class field information can be extracted from all non-string, non-array objects found on the gc heap

noahfalk commented on src/coreclr/vm/proftoeeinterfaceimpl.cpp:
```diff
@@ -6846,11 +6850,18 @@ HRESULT ProfToEEInterfaceImpl::SuspendRuntime()
+    {
+        LOG((LF_CORPROF,
+             LL_ERROR,
+             "**PROF: ERROR: Returning CORPROF_E_UNSUPPORTED_CALL_SEQUENCE due to illegal asynchronous profiler call\n"));
```
Oops I had a copy/paste bug in the logging text I suggested earlier :(
```suggestion
             "**PROF: ERROR: Returning CORPROF_E_UNSUPPORTED_CALL_SEQUENCE due to illegal gc-triggers profiler call inside a no-trigger callback\n"));
```
Same thing the other places this check happens.

noahfalk reviewed (approved):
A couple minor suggestions inline, overall looks good! I'll also respond back to some of the existing comments in just a moment.

mdh1418 committed:
Update EnumerateGCHeapObjects usage description

mdh1418 committed:
Cleanup

mdh1418 committed:
Add validation to extract class fields from objects

mdh1418 commented on src/coreclr/inc/corprof.idl:
```diff
@@ -4308,6 +4312,37 @@ interface ICorProfilerInfo14 : ICorProfilerInfo13
+     * EnumerateGCHeapObjects is a method that iterates over each object in the GC heap.
+     * For each object, it invokes the provided callback function which should return a bool
+     * indicating whether or not enumeration should continue.
+     * The runtime must be suspended in order to enumerate objects on the GC heap.
```
Updated the guidance!

mdh1418 commented:
The dotnet-linker-tests failure should be fixed by https://github.com/dotnet/runtime/pull/104606
